### PR TITLE
Functional Widgets and Render Middleware

### PR DESCRIPTION
**Pull request by agubler**
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code has been formatted with [`prettier`](https://prettier.io/) as per the [readme code style guidelines](./../#code-style)
* [x] Unit or Functional tests are included in the PR

**Description:**

The basic implementation to support functional widgets and render middleware. Includes one core middleware `invalidator` (open to renaming suggestions).

Provides a single function `create` that returns a factory that can be used to create Dojo widgets and render middleware. 

**Note:** Currently `create` is in the `tsx` module in `widget-core`, this is going to change as the feature is built on.

##### Creates a simple widget with a properties interface:

```ts
const { create } from '@dojo/framework/widget-core/tsx';

const render = create().properties<{ foo: string }>();
const MyWidget = render(() => v('div'));
```

##### Creates a render middleware with a properties interface:

```ts
const { create } from '@dojo/framework/widget-core/tsx';

const middleware = create().properties<{ bar: string }>();
const myMiddleware = middleware(() => {
    return {
        get() {},
        set() {}
    };
});
```

##### Creates a widget with middleware that influences the property typings:

```ts
const { create } from '@dojo/framework/widget-core/tsx';

import myMiddleware from './myMiddleware';

const render = create({ myMiddleware }).properties<{ foo: string }>();
const MyWidget = render(({ properties, middleware }) => {
    // can access the injected middleware
    middleware.myMiddleware.get();
    middleware.myMiddleware.set();
    // properties API defined by the widget and the injected middleware
    return v('div', [properties.foo, properties.bar]);
});
```

The change includes support for calling a widget directly, calling the widget directly has the added advantages to be able to infer complex properties such as a render property, but is also backward compatible for use with `w()`.

**Example Usages:**

```tsx
// Supports the existing w() pragma
w(MyWidget, { foo: 'bar' });

// Support being called directly
MyWidget({ foo: 'bar' });

// Supports tsx
<MyWidget foo="bar" />
```

Related to #349

### CODE REVIEW

File: src/widget-core/Registry.ts [hunk 1]
```diff
@@ -87,16 +95,23 @@ export interface RegistryInterface {
...
+	return isWidgetBaseConstructor(item) || isWidgetFunction(item);
+}
+
 export function isWidgetConstructorDefaultExport<T>(item: any): item is ESMDefaultWidgetBase<T> {
 	return Boolean(
-		item &&
-			item.hasOwnProperty('__esModule') &&
-			item.hasOwnProperty('default') &&
-			isWidgetBaseConstructor(item.default)
+		item && item.hasOwnProperty('__esModule') && item.hasOwnProperty('default') && isWidget(item.default)
```
Comment (matt-gadd) [inline src/widget-core/Registry.ts]: I think for code splitting this needs to also check whether it's a widget factory.

File: src/widget-core/middleware/core.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import { create } from '../tsx';
+import { getInvalidator } from '../vdom';
+
+const createFactory = create();
+
+export const invalidator = createFactory(({ id }) => {
```
Comment (matt-gadd) [inline src/widget-core/middleware/core.ts]: starting to wonder whether we should put these base ones in vdom itself? then we don't have to export the function itself?

File: src/widget-core/tsx.ts [hunk 3]
```diff
@@ -54,3 +64,51 @@ export function tsx(tag: any, properties = {}, ...children: any[]): DNode {
...
+		};
+	};
+	factory.isFactory = true;
+	return factory;
+}
+
+export function create<T extends MiddlewareMap<any>, MiddlewareProps = ReturnType<T[keyof T]>['properties']>(
+	middlewares: T = {} as T
+) {
+	function properties<Props extends {}>() {
```
Comment (sbinge) [inline src/widget-core/tsx.ts]: Would this read better as something similar to `withProperties()`? Thinking about it from a perspective of making the create API more fluent, e.g.:

```tsx
create().withProperties<{]>();
```

File: src/widget-core/middleware/core.ts (hunk 2 above)
Comment (agubler) [reply src/widget-core/middleware/core.ts]: I think this is a good idea, will have more control in vdom itself also.

File: src/widget-core/tsx.ts (hunk 3 above)
Comment (agubler) [reply src/widget-core/tsx.ts]: I was trying to reduce the number of characters but when with `properties`, I don't actually mind the `with` prefix tho for it's fluency. @matt-gadd any thoughts?

**Comment by agubler** [comment]
@sbinge we do have an implementation of a js-framework-benchmark that I have been upgrading and running using functional widgets. The performance is similar and the memory usage slightly less. I will do a side by side comparison of the two different implementation and post onto this issue.

File: src/widget-core/Registry.ts (hunk 1 above)
Comment (agubler) [reply src/widget-core/Registry.ts]: Updated

File: src/widget-core/middleware/core.ts (hunk 2 above)
Comment (agubler) [reply src/widget-core/middleware/core.ts]: Done

**Comment by agubler** [comment]
@sbinge running two implementations of the js-framework-benchmark, one with classes, one with functional widgets but otherwise identical corroborates the performance these characteristics.

Interactive_Results

File: src/widget-core/tsx.ts (hunk 3 above)
Comment (sbinge) [reply src/widget-core/tsx.ts]: There's another subtle comprehension hiccup here - this method is describing the property interface as part of the factory creation, not widget creation nor rendering invocation.

Reading it as it currently stands, a user could potentially confuse it with passing actual property values as part of a 'create a widget` call; something that `withProperties` or `hasProperties` doesn't help with to resolve.

What about something like `withAPI()`?

Comment (agubler) [reply src/widget-core/tsx.ts]: My issue with that is that in the future, when typescript generics catch up we’d like to explore the possibility of using the function to define default properties which will leave the external property API with optional properties but enable referencing them from within widget knowing that although the external API is optional the internal API is not. For the moment typescript would prevent a user passing anything to the function as there are no arguments current.

Additionally the API of a widget is not just properties, it’s also children which again in the future is something that we’d consider exposing for the user to define perhaps with a further `.children` API.

Comment (sbinge) [reply src/widget-core/tsx.ts]: Makes sense.

**Pull request merged**